### PR TITLE
Implement promo discount codes that can be specified in percentage or absolute terms

### LINKED ISSUE #1136: Promo codes
connects to big-neon/bigneon#45
@zmelamed do we need to maintain the face value for the reports?

### PR DESCRIPTION
**Pull request by philipr-za**
### Closes Issues:
Closes https://github.com/big-neon/bn-api/issues/1136
connects to https://github.com/big-neon/bigneon/issues/45

### Description:
This PR builds on the discount and access code implementation from https://github.com/big-neon/bn-api/issues/446 to bring the functionality in line with current UX requirements.

Discount codes can now be created with either a percentage discount OR an absolute discount (not both, this is validated). If a discount code is used during a cart update the discount will be applied by adding a Discount order item with a negative price. The discount amount is capped at maximum the price of the ticket being discounted so at most a ticket can be discounted to being free.

A change in paradigm from the initial discount code implementation is that initially the code was written so that a `Use` of a code corresponds to an order_item i.e. an order_item for 4 tickets using a given code would be counted as one use of that code. This PR changes the measurement of uses to be based on quantity of tickets purchased with that code. So an order_item for 4 tickets is 4 uses of the discount code. 

The GET /redemption_code/{code} endpoint now includes an `available` field which states how many overall uses of that code are still available.

To do on the front-end:
- Handle the new Discount order item when rendering the cart
- Allow for creating and managing discount codes on events including the two possible types of discounts (percentage and absolute)
- When querying discount codes to apply to a cart read the total number available (found in the new `available` field) and the per-user limit and display the limits
- When querying redemption codes via GET /redemption_code/{code} endpoint add the event_id as a query string to limit returned results i.e. GET /redemption_code/{code}?event_id=0000-000000-0000-0000

### Migrations
- 20190318122751_add_new_index_and_percentage_to_codes

### Environment Variables
 * No change

### CODE REVIEW

File: db/src/models/codes.rs [hunk 1]
```diff
@@ -87,6 +98,44 @@ impl Code {
...
+        let used: Vec<i64> = order_items::table
+            .left_join(orders::table.on(order_items::order_id.eq(orders::id)))
+            .filter(order_items::order_id.ne(order_id_to_exclude.unwrap_or(Uuid::nil())))
+            .filter(order_items::code_id.eq(code_id))
+            .filter(
+                orders::expires_at
+                    .gt(dsl::now.nullable())
+                    .or(orders::status.eq(OrderStatus::Paid)),
+            )
+            .select(order_items::quantity)
```
Comment (Timo614) [inline db/src/models/codes.rs]: Probably not a huge deal but may be worth subtracting out the `order_items::refunded_quantity` so we only count the quantity that the user ended up with. Probably an edge case though as not many people who ask for refunds will turn around and purchase more for a given event.

File: db/migrations/20190318122751_add_new_index_and_percentage_to_codes/down.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
+DROP INDEX index_redemption_code_event_id;
+
+UPDATE codes
+SET max_uses = 9223372036854775807 where max_uses = 0; -- previously no limit
```
[the remark this answers is not in the document]
Comment (philipr-za) [reply db/migrations/20190318122751_add_new_index_and_percentage_to_codes/down.sql]: So there was a constraint on max_uses that checked that `max_uses>0` so the front end can't have used it until I dropped that constraint in the associated `up.sql` and you can see me re-add the constraint below.

File: db/src/models/codes.rs (hunk 1 above)
Comment (philipr-za) [reply db/src/models/codes.rs]: Good point, added in a clause only to count uses where the refunded_quantity is zero.

File: db/src/models/codes.rs [hunk 3]
```diff
@@ -87,6 +98,44 @@ impl Code {
...
+        Ok(CodeAvailability { code, available })
+    }
+
+    pub fn find_number_of_uses(
+        code_id: Uuid,
+        order_id_to_exclude: Option<Uuid>,
+        conn: &PgConnection,
+    ) -> Result<i64, DatabaseError> {
+        let used: Vec<i64> = order_items::table
+            .left_join(orders::table.on(order_items::order_id.eq(orders::id)))
```
[the remark this answers is not in the document]
Comment (philipr-za) [reply db/src/models/codes.rs]: Done

File: db/src/models/codes.rs [hunk 4]
```diff
@@ -87,6 +98,44 @@ impl Code {
             )
     }
 
+    pub fn find_by_redemption_code_with_availability(
```
[the remark this answers is not in the document]
Comment (philipr-za) [reply db/src/models/codes.rs]: Had to refactor a little in a couple places but I have removed `find_by_redemption_code`

File: db/src/models/codes.rs [hunk 5]
```diff
@@ -87,6 +98,44 @@ impl Code {
...
+        redemption_code: &str,
+        conn: &PgConnection,
+    ) -> Result<CodeAvailability, DatabaseError> {
```
[the remark this answers is not in the document]
Comment (philipr-za) [reply db/src/models/codes.rs]: Oh yes, I forgot about that. Added in the ability to provide the event_id in the query string for GET /redemption_code/{code} and an Option<Uuid> in this function to cater for searching with an event_id filter.

File: db/src/models/codes.rs [hunk 6]
```diff
@@ -87,6 +98,44 @@ impl Code {
...
+                ErrorCode::QueryError,
+                "Could not load code with that redeem code",
+            )?;
+
+        let available = code.max_uses - Code::find_number_of_uses(code.id, None, conn)?;
+
+        Ok(CodeAvailability { code, available })
+    }
+
+    pub fn find_number_of_uses(
```
[the remark this answers is not in the document]
Comment (philipr-za) [reply db/src/models/codes.rs]: Implemented a test for this function. I checked the following cases:

Add some tickets to the cart = code uses not available
Expire the cart = code uses become available again
Buy tickets and expire the cart = code uses not available
Refund bought tickets = code use available again

**Pull request merged**